osu.Game/Screens/Select/SongSelect.cs

```diff
@@ -448,8 +448,10 @@ void run()
                 {
                     Mods.Value = Array.Empty<Mod>();
 
-                    // required to return once in order to have the carousel in a good state.
-                    // if the ruleset changed, the rest of the selection update will happen via updateSelectedRuleset.
+                    // the ruleset transfer may cause a deselection of the current beatmap (due to incompatibility).
+                    // this can happen via Carousel.FlushPendingFilterOperations().
+                    // to ensure a good state, re-transfer no-debounce values.
```

Comment:
I *think* I understand the fix (took me a solid half an hour of staring at the control flow, but that's probably just my brain leaking), but this comment reads a bit backwards even in light of the PR description? Everything in it is *correct*, the transfer may cause deselection, but in this particular failure case it's the *lack* of deselection which causes the `run` delegate to not fire the second time, right? Again, might just be me, but this reads a little like it was the other way around.

Reply:
I agree this reads pretty badly still. Will refactor accordingly...

Reply:
@bdach I refactored this with a slightly different approach that should hopefully be more self-explanatory.

A next step forward may be reversing the flow such that `workingBeatmapChanged` just calls `performUpdateSelected` and the carousel select logic is moved to `performUpdateSelected`, but I'm not ready to take that on just yet..